app/grpc/gasestimation/gas_estimator.go

```diff
@@ -105,144 +88,120 @@ func (s *gasEstimatorServer) EstimateGasPriceAndUsage(ctx context.Context, reque
 	}, nil
 }
 
+// gasPriceEstimationThreshold the threshold of mempool transactions to
+// estimate the gas price.
+var gasPriceEstimationThreshold = 0.70
+
 // estimateGasPrice takes a transaction priority and estimates the gas price based
-// on the gas prices of the transactions in the last five blocks.
-// If no transaction is found in the last five blocks, return the network
-// min gas price.
-// It's up to the light client to set the gas price in this case
-// to the minimum gas price set by that node.
+// on the gas prices of the transactions in the mempool.
+// If the mempool transactions can't fill more than 70% of the block, the min gas
+// price is returned.
 func (s *gasEstimatorServer) estimateGasPrice(ctx context.Context, priority TxPriority) (float64, error) {
-	status, err := s.clientCtx.Client.Status(ctx)
+	// using -1 to return all the transactions.
+	limit := -1
+	txsResp, err := s.mempoolClient.UnconfirmedTxs(ctx, &limit)
 	if err != nil {
 		return 0, err
 	}
-	latestHeight := status.SyncInfo.LatestBlockHeight
-	page := 1
-	txSearchResult, err := s.clientCtx.Client.TxSearch(
-		ctx,
-		lastFiveBlocksTransactionsQuery(latestHeight),
-		false,
-		&page,
-		&numberOfTransactionsPerPage,
-		"asc",
-	)
+	if txsResp.TotalBytes < int64(appconsts.DefaultMaxBytes*gasPriceEstimationThreshold) {
+		return appconsts.DefaultMinGasPrice, nil
+	}
```

Comment:
[question] this doesn't account for padding in a data square. The txs can take up < 70% of the bytes of DefaultMaxBytes and still take up > 100% of the data square.

[blocking] appconsts.DefaultMaxBytes shouldn't be used here because it's still ~1.9 MiB. On all networks, governance has modified it to be 8 MiB so this won't work as expected.

Reply:
> [question] this doesn't account for padding in a data square. The txs can take up < 70% of the bytes of DefaultMaxBytes and still take up > 100% of the data square.

Yes, that's true. The issue if we want to account for the padding is needing to build a square and that's too expensive to do on every estimation call. I guess this value is enough since it covers at least 70% of the block or more.

> [blocking] appconsts.DefaultMaxBytes shouldn't be used here because it's still ~1.9 MiB. On all networks, governance has modified it to be 8 MiB so this won't work as expected.

good catch 👍 

6460fd1ad74985817d945cb762743f4c47c98330